File: C++/InsertionSort.cpp
#include <iostream>
#include <ctime>
#include <cassert>
#include "SortTestHelper.h"
#include "SelectionSort.h"

using namespace std;

/* 插入排序 */
template<typename T>
void insertionSort(T arr[], int n) {

    for (int i = 1; i < n; i++) {
        // 寻找元素arr[i]合适的插入位置
        T e = arr[i];
        int j;
        for (j = i; j > 0 && arr[j-1] > e; j--) {
            arr[j] = arr[j-1];
        }
        arr[j] = e;
    }
}

template<typename T>
void insertionSort(T arr[], int l, int r) {
    int n = r - l + 1;
    for (int i = l; i <= r; i++) {
        T e = arr[i - l];
        int j;
        for (j = i; j > 0 && arr[j-l-1] > e; j--) {
            arr[j-l] = arr[j-l-1];
        }
        arr[j-l] = e;
    }
}

int main() {
    int n = 100000;
    int *arr = SortTestHelper::generateNearlyOrderedArray(n, 10);
    int *arr2 = SortTestHelper::copyIntArray(arr, n);

    SortTestHelper::testSort("Insertion Sort", insertionSort, arr, n);
    SortTestHelper::testSort("Selection Sort", selectionSort, arr2, n);

    delete[] arr;
    delete[] arr2;

    cout << "Hello Wsx" << endl;
    return 0;
}

File: JavaScript/algorithms/sorting/insertion-sort/InsertionSort.js
/**
 * 插入排序算法
 * 最优时间复杂度 O(n)
 * 平均时间复杂度 O(n^2)
 * 最坏时间复杂度 O(n^2)
 * 占用内存
 * 是否稳定 是
 */
class InsertionSort {
  sort(originalArray) {
    let array = [...originalArray];

    // 从第一个元素开始， 这个元素可以认为已经被排序
    for (let i = 1; i < array.length; i++) {

      // 取出下一个元素
      let temp = array[i];

      // 从已经排序的元素序列中 从后向前扫描
      // 如果这个元素大于新元素 则交换位置，将该元素移到下一位置
      for (let j = i - 1; j >= 0 && temp < array[j]; j--) {
        array[j + 1] = array[j];
        array[j] = temp;
      }
    }

    return array;
  }
}

let array = [6, 1, 2, 4, 5, 6, 8, 9, 3, 0];
let insertion = new InsertionSort();
let res = insertion.sort(array);
console.log(res);
